File: Data Structures & Algorithms/palindrome-partitioning/submission-20.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = [] 

        def dfs(start, cur): 
            if start == len(s): 
                res.append(cur.copy()) 
                return 

            for i in range(start, len(s)): 
                partition = s[start: i + 1] 
                if self.isValid(partition): 
                    cur.append(partition) 
                    dfs(i + 1, cur) 
                    cur.pop() 

        dfs(0, []) 
        return res




    def isValid(self, s): 
        if len(s) == 1: 
            return True 

        l = 0 
        r = len(s) - 1 

        while l < r: 
            if s[l] != s[r]: 
                return False 

            l += 1 
            r -= 1 

        return True

File: Data Structures & Algorithms/subsets/submission-20.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) { 
        const res = [] 
        this.rec(0, nums, [], res) 
        return res
    } 


    rec(start, nums, part, res) {
        res.push([...part]) 

        for (let i = start; i < nums.length; i ++) {
            part.push(nums[i]) 
            this.rec(i + 1, nums, part, res) 
            part.pop()
        }
    }
}
